### PR TITLE
[SPARK-37173][SQL] SparkGetFunctionOperation return builtin function only once

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
According to https://github.com/apache/spark/pull/25252/files#r738489764, if we use wild pattern, it will return too much rows.

In this pr we return common builtin functions only once

### Why are the changes needed?
Improve performance

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
WIP

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @wangyum @juliuszsompolski

**Comment by juliuszsompolski** [comment]
In https://github.com/apache/spark/pull/25252 @dongjoon-hyun suggests to make it under a legacy feature flag, since it's behaviour changing.

**Comment by dongjoon-hyun** [comment]
Thank you, @juliuszsompolski

**Comment by AngersZhuuuu** [comment]
> In #25252 @dongjoon-hyun suggests to make it under a legacy feature flag, since it's behaviour changing.

Sure， will start work on this.

**Comment by AngersZhuuuu** [comment]
> legacy feature flag

For legacy flag, should we must use `spark.sql.legacy.xxx` and default value is false?

**Comment by juliuszsompolski** [comment]
cc @bogdanghit

**Comment by AngersZhuuuu** [comment]
GA failed caused by oom

**Comment by dongjoon-hyun** [comment]
It's up to our decision, but it's safe to keep the existing behavior first and to switch it at next version because it gives the users a chance to try this new behavior.
> For legacy flag, should we must use spark.sql.legacy.xxx and default value is false?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1158,6 +1158,14 @@ object SQLConf {
     .intConf
     .createWithDefault(200)
 
+  val THRIFTSERVER_SEPARATE_DISPLAY_SYSTEM_FUNCTION =
+    buildConf("spark.sql.thriftserver.separateDisplaySystemFunctions")
+      .doc("When true, Spark Thrift Server will return databases' system functions metadata " +
+        "only once, and Spark will change function schema as `SYSTEM`.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1164]: Does Apache Spark use `SYSTEM` before?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala [hunk 2]
```diff
@@ -80,8 +83,20 @@ private[hive] class SparkGetFunctionsOperation(
       parentSession.getUsername)
 
     try {
+      val separateDisplaySystemFunctions =
+        sqlContext.conf.getConf(SQLConf.THRIFTSERVER_SEPARATE_DISPLAY_SYSTEM_FUNCTION)
+      var matchedBuiltInFunctions = if (separateDisplaySystemFunctions && functionPattern == "*"
+        && matchingDbs.nonEmpty) {
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala:89]: indentation?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala [hunk 3]
```diff
@@ -80,8 +83,20 @@ private[hive] class SparkGetFunctionsOperation(
...
+        FunctionRegistry.functionSet ++ TableFunctionRegistry.functionSet
+      } else {
+        Set.empty[FunctionIdentifier]
+      }
       matchingDbs.foreach { db =>
         catalog.listFunctions(db, functionPattern).foreach {
+          case (funcIdentifier, "SYSTEM") if separateDisplaySystemFunctions =>
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala:96]: Please declare a constant variable and use it.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala [hunk 4]
```diff
@@ -94,6 +109,17 @@ private[hive] class SparkGetFunctionsOperation(
             rowSet.addRow(rowData);
         }
       }
+      matchedBuiltInFunctions.foreach { functionIdentifier =>
+        val info = catalog.lookupFunctionInfo(functionIdentifier)
+        val rowData = Array[AnyRef](
+          DEFAULT_HIVE_CATALOG, // FUNCTION_CAT
+          "SYSTEM", // FUNCTION_SCHEM
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala:116]: ditto.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1164]: > Does Apache Spark use `SYSTEM` before in other places?
sessionCatalog.listFunctions()
```
    // so there can be duplicates.
    functions.map {
      case f if FunctionRegistry.functionSet.contains(f) => (f, "SYSTEM")
      case f if TableFunctionRegistry.functionSet.contains(f) => (f, "SYSTEM")
      case f => (f, "USER")
    }.distinct
```

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala]: DOne

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala]: Done

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1164]: Got it~

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 5]
```diff
@@ -236,6 +240,23 @@ class SparkMetadataOperationSuite extends HiveThriftServer2TestBase {
       checkResult(metaData.getFunctions(null, "default", "shift*"),
         Seq("shiftleft", "shiftright", "shiftrightunsigned"))
       checkResult(metaData.getFunctions(null, "default", "upPer"), Seq("upper"))
+
+      statement.execute(s"SET ${SQLConf.THRIFTSERVER_SEPARATE_DISPLAY_SYSTEM_FUNCTION.key}=true")
```
Comment (bogdanghit) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:244]: Can we add a test with two schemas and run an unfiltered getFunctions call to show that previously we'd see duplicates, whereas now the functions are unique?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -1158,6 +1158,14 @@ object SQLConf {
     .intConf
     .createWithDefault(200)
 
+  val THRIFTSERVER_SEPARATE_DISPLAY_SYSTEM_FUNCTION =
+    buildConf("spark.sql.thriftserver.separateDisplaySystemFunctions")
+      .doc("When true, Spark Thrift Server will return databases' system functions metadata " +
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1163]: Done

File: docs/sql-migration-guide.md [hunk 7]
```diff
@@ -54,6 +54,8 @@ license: |
 
   - Since Spark 3.3, nulls are written as empty strings in CSV data source by default. In Spark 3.2 or earlier, nulls were written as empty strings as quoted empty strings, `""`. To restore the previous behavior, set `nullValue` to `""`.
 
+  - Since Spark 3.3, Spark Thrift Server will return databases' system functions metadata only once, and Spark will change function schema as `SYSTEM`. In Spark 3.2 or earlier, Spark Thrift Server will return system functions metadata for all databases. To restore the behavior before Spark 3.3, yo you can set `spark.sql.thriftserver.separateDisplaySystemFunctions` to `false`.
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply docs/sql-migration-guide.md:57]: Done

**Comment by juliuszsompolski** [comment]
> It's up to our decision, but it's safe to keep the existing behavior first and to switch it at next version because it gives the users a chance to try this new behavior.

My 2cents: working with various partners and vendors of BI tools, we found GetFunctions to rarely be used at all, and when we found it used, it was in the context of the current behaviour causing trouble (running to slow, causing UI freezes because of trying to render the humongous list of duplicated functions). I am not aware of any tool depending on the current behaviour.
Note that before GetTables / GetSchemas / GetColumns gotten implemented in Spark 3.0, this was all throwing wrong results or errors with Spark, and it has not been escalated for a long time. Only recently BI vendors are picking up serious interest in developing connectors that take advantage of these functions, and this has been reported as something that is an unexpected and unwanted behaviour.

**Comment by AngersZhuuuu** [comment]
> > It's up to our decision, but it's safe to keep the existing behavior first and to switch it at next version because it gives the users a chance to try this new behavior.
> 
> My 2cents: working with various partners and vendors of BI tools, we found GetFunctions to rarely be used at all, and when we found it used, it was in the context of the current behaviour causing trouble (running to slow, causing UI freezes because of trying to render the humongous list of duplicated functions). I am not aware of any tool depending on the current behaviour. Note that before GetTables / GetSchemas / GetColumns gotten implemented in Spark 3.0, this was all throwing wrong results or errors with Spark, and it has not been escalated for a long time. Only recently BI vendors are picking up serious interest in developing connectors that take advantage of these functions, and this has been reported as something that is an unexpected and unwanted behaviour.

We use HUE for adhoc before, to support use hue in spark's thrift server. we also changed a lot..

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala]: Done

**Review by bogdanghit — APPROVED** [review]
LGTM, thanks @AngersZhuuuu

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 5 above)
Comment (bogdanghit) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala]: Great, thanks!

**Comment by AngersZhuuuu** [comment]
gentle ping @dongjoon-hyun  WDYT of current code?

**Comment by AngersZhuuuu** [comment]
Any more suggestion? also cc @wangyum

**Comment by AngersZhuuuu** [comment]
Any more suggestion?

**Comment by bogdanghit** [comment]
> Any more suggestion?

It looks good to me. @dongjoon-hyun WDYT?

**Comment by bogdanghit** [comment]
Gentle nudge @AngersZhuuuu @dongjoon-hyun @wangyum.

**Comment by bogdanghit** [comment]
@dongjoon-hyun what should still be done to push this through?

**Comment by srowen** [comment]
Would the existing behavior ever be desirable? sounds like more of a bug?

**Comment by bogdanghit** [comment]
@srowen that was my initial thought as well, but there are concerns it may be a breaking change because of the different result format

**Pull request closed without merging**